sympy/core/mul.py

```diff
@@ -195,6 +195,10 @@ def flatten(cls, seq):
                                 bargs.insert(0, ar)
                             bargs = [Add._from_args(bargs)]
                             rv = bargs, [], None
+            if a.is_commutative and b.is_commutative:
+                if ((a.is_finite == False and b.is_zero) or
+                        (b.is_finite == False and a.is_zero)):
+                    rv = [S.NaN], [], None
```

Comment:
I am not following completely. If I don't set the value of `rv`, the flow will just move forward like earlier. Am I missing something?